src/lib/verifier/prod.ml

```diff
@@ -313,25 +313,26 @@ let create ~logger ~proof_level ~pids ~conf_dir : t Deferred.t =
                   [%log info] "verifier successfully got sigkill"
                     ~metadata:[("verifier_pid", `Int (Pid.to_int pid))] )
         in
-        let new_worker =
-          let%bind exit_metadata =
-            match%map wait_safe process with
-            | Ok res ->
-                [ ( "exit_status"
-                  , `String (Unix.Exit_or_signal.to_string_hum res) ) ]
-            | Error err ->
-                [ ("exit_status", `String "Unknown: wait threw an error")
-                ; ("exn", Error_json.error_to_yojson err) ]
-          in
-          [%log info] "verifier successfully stopped"
-            ~metadata:
-              ( ("verifier_pid", `Int (Process.pid process |> Pid.to_int))
-              :: exit_metadata ) ;
-          Child_processes.Termination.remove pids pid ;
-          let%map worker = create_worker () in
-          on_worker worker ; worker
-        in
-        worker_ref := new_worker )
+        let new_worker = Ivar.create () in
+        worker_ref := new_worker ;
+        don't_wait_for
+        @@ let%bind exit_metadata =
+             match%map wait_safe process with
```

Comment:
Looked at the logs for `seed-2` on `finalfinal2`, which is currently in this state. This appears to be the likely lock point.

To summarise my hypothesis briefly:
* this calls `Process.wait`
* `Process.wait` forces a lazy call to wait for the child process
* since this can be scheduled way later, the process may have already gone away, and the wait hangs forever

If this is the cause, I think we need to
* call `Process.wait` immediately on process creation, storing the `Async.Deferred.t` in a value to be `let%bind`ed when we actually need it
  - this should be wrapped in a synchronous `try ... with _ -> ...` so that we don't get WNOHANG exceptions
  - this should probably be at or around the `Daemon started process` log message, otherwise it's too early
* bind the value here to find out the exit status
  - in a `Deferred.try_with`, since it may raise

The situation should already be improved by this PR, but if the process may exit before the scheduler gets to this `wait_safe` call (which I suspect it can) then we'll need to do this full fix.